internal/resolver/delegatingresolver/delegatingresolver.go

```diff
@@ -0,0 +1,195 @@
+/*
+ *
+ * Copyright 2024 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package delegatingresolver implements the default resolver that creates child
+// resolvers to resolver targetURI as well as proxy adress.
+
+package delegatingresolver
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"sync"
+
+	"google.golang.org/grpc/attributes"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/serviceconfig"
+)
+
+type delegatingResolver struct {
+	target         resolver.Target
+	cc             resolver.ClientConn
+	isProxy        bool
+	targetResolver resolver.Resolver
+	ProxyResolver  resolver.Resolver
+	targetAddrs    []resolver.Address
+	proxyAddrs     []resolver.Address
+	proxyURL       *url.URL
+	mu             sync.Mutex // Protects the state below
+
+}
+
+type innerClientConn struct {
+	*delegatingResolver
```

Comment:
Why do you have to embed a pointer to the parent here? Embedding structs can lead to some strange behaviors. This needs to be well understood before embarking.

Would embedding the `resolver.Resolver` interface work instead?

Reply:
Maybe you can instead store a `parent *delegatingResolver` field in this type, and in the `UpdateState` method, call a new method on the `delegatingResolver` that handles updates and returns `resolver.State` that the `innerClientConn` passes to `icc.cc.UpdateState`.